### PR TITLE
callable-types: Refactor and add fixer

### PR DESCRIPTION
**Pull request by ajafff**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:
[bugfix] `callable-types`: suggest correct fix for interfaces with type arguments
[new-fixer] `callable-types`

### CODE REVIEW

File: src/rules/callableTypesRule.ts [hunk 1]
```diff
@@ -29,76 +30,68 @@ export class Rule extends Lint.Rules.AbstractRule {
...
         }
-    }
+        return ts.forEachChild(node, cb);
+    });
 }
 
 /** True if there is no supertype or if the supertype is `Function`. */
-function noSupertype(heritageClauses: ts.NodeArray<ts.HeritageClause> | undefined): boolean {
-    if (!heritageClauses) {
+function noSupertype(node: ts.InterfaceDeclaration): boolean {
```
Comment (andy-hanson) [inline src/rules/callableTypesRule.ts]: Why did you change this to take `node` even when you're only using the `.heritageClauses`?

File: test/rules/callable-types/test.ts.lint [hunk 2]
```diff
@@ -1,23 +1,36 @@
...
+
+interface L<T> {
+    (): T
+    ~~~~~ [Interface has only a call signature — use `type L<T> = () => T` instead.]
 }
 
 type T = {
     (): void;
-    ~~~~~~~~~ [Type literal has only a call signature — use `() => void` instead.]
+    ~~~~~~~~~ [Type literal has only a call signature — use `() => void;` instead.]
```
Comment (andy-hanson) [inline test/rules/callable-types/test.ts.lint]: Don't think this should have a semicolon. A type literal may not necessarily appear at the end of a statement, could be `function f(x: { (): void }): void;`, shouldn't fix to `function f(x: () => void;): void;`.

File: src/rules/callableTypesRule.ts (hunk 1 above)
Comment (ajafff) [inline src/rules/callableTypesRule.ts]: it makes the signature shorter and the intention of the function clearer. Of course it could be destructured directly in the parameter, but I doubt this would make a difference

File: test/rules/callable-types/test.ts.lint (hunk 2 above)
Comment (ajafff) [inline test/rules/callable-types/test.ts.lint]: it actually handles that already. The only exception is on type aliases. But since a TypeAlias should always end with a semicolon, I guess I can always remove the semicolon from the call signature

File: src/rules/callableTypesRule.ts [hunk 3]
```diff
@@ -29,76 +30,68 @@ export class Rule extends Lint.Rules.AbstractRule {
...
+                          parent: ts.InterfaceDeclaration | ts.TypeLiteralNode,
+                          sourceFile: ts.SourceFile): string {
+    const start = call.getStart(sourceFile);
+    const colonPos = call.type!.pos - 1 - start;
+    const text = sourceFile.text.substring(start, call.end);
+    let res = text.substr(0, colonPos) + " =>" + text.substr(colonPos + 1);
+
+    if (parent.kind === ts.SyntaxKind.InterfaceDeclaration) {
+        if (parent.typeParameters !== undefined) {
+            res = `type${sourceFile.text.substring(parent.name.pos, parent.typeParameters.end + 1)} = ${res}`;
```
Comment (andy-hanson) [inline src/rules/callableTypesRule.ts]: Just return here instead of using `res =`.
